substratevm/DebugInfo.md

```diff
@@ -90,12 +165,346 @@ The currently implemented features include:
+_objhdr which includes a single field caled `hub` whose type is
+`java.lang.Class *` i.e. it is a pointer to the obejct's class.
+
+The ptype command can be used to print details of a specific type. Note
+that the java type name must be specified in quotes because to escape the
+embedded `.` characters.
+
+```
+(gdb) ptype 'java.lang.String'
+type = class java.lang.String : public java.lang.Object {
+  private:
+    byte [] *value;
+    int hash;
+    byte coder;
+
+  public:
+    void String(byte [] *);
+    void String(char [] *);
+    void String(byte [] *, java.lang.String *);
+    . . .
+    char charAt(int);
+    . . .
+    java.lang.String * concat(java.lang.String *);
+    . . .
+}
+```
+
+The print command can be used to print the contents of a referenced object
+field by field. Note how a cast is used to convert a raw memory address to
+a reference for a specific Java type.
+
+```
+(gdb) print *('java.lang.String' *) 0x7ffff7c01060
+$1 = {
+  <java.lang.Object> = {
+    <_objhdr> = {
+      hub = 0x90cb58
+    }, <No data fields>}, 
+  members of java.lang.String:
+  value = 0x7ffff7c011a0,
+  hash = 0,
+  coder = 0 '\000'
+}
+```
+
+The hub field in the object header is actually a reference of Java type
+`java.lang.Class`. Note that the field is typed by gdb using a pointer
+to the underlying C++ class (layout) type.
+
+```
+(gdb) ptype _objhdr
+type = struct _objhdr {
+    java.lang.Class *hub;
+}
+```
+
+Given an address that might be an object reference it is possible to
+verify that case and identify the object's type by printing the
+contents of the String referenced from the hub's name field.  First
+the value is cast to an object reference. Then a path expression is
+used to dereference through the the hub field and the hub's name field
+to the `byte[]` value array located in the name String.
+
+```
+(gdb) print/x ((_objhdr *)$rdi)
+$2 = 0x7ffff7c01028
+(gdb) print *$2->hub->name->value
+$3 = {
+  <_arrhdrB> = {
+    hub = 0x90e4b8,
+    len = 19,
+    idHash = 3759493
+  }, 
+  members of byte []:
+  data = 0x904798 "[Ljava.lang.String;"
+}
+```
+
+The value in register rdx is obviously a reference to a String array.
+Casting it to this type shows it has length 1.
+
+```
+(gdb) print *('java.lang.String[]' *)$rdi
+$4 = {
+  <_arrhdrA> = {
+    hub = 0x906a78,
+    len = 2,
+    idHash = 0
+  }, 
+  members of java.lang.String[]:
+  data = 0x7ffff7c01038
+}
+```
+
+A simpler command which allows just the name of the hub object to be
+printed is as follows:
+
+```
+(gdb) x/s $2->hub->name->value->data
+798:	"[Ljava.lang.String;"
+```
+
+Indeed it is useful to define a gdb command `hubname` to execute this
+operation on an arbitrary input argument
+
+```
+command hubname
+  x/s (('java.lang.Class' *)((long)((('java.lang.Object' *)($arg0))->hub) & ~0x7L))->name->value->data
+end
+
+(gdb) hubname $2
+0x904798:	"[Ljava.lang.String;"
+```
+
+Notice that the `hubname` command also masks out the low 3 flag bits in
```

Comment:
With proper implementation the manual masking of the flag bits should not be necessary. It is possible to implement the masking of the three bits in the dwarf-type of the hub object. The `__hub__` member in the above examples https://github.com/oracle/graal/pull/3046/files#r562532653 is of type `<0x002d64e0>`. Here is the definition of this dwarf-type:
```
< 1><0x002d64e0>    DW_TAG_pointer_type
                      DW_AT_type                  <0x002d64e6>
                      DW_AT_byte_size             0x00000004
< 1><0x002d64e6>    DW_TAG_structure_type
                      DW_AT_name                  _z_.java.lang.Class
                      DW_AT_byte_size             0x00000078
                      DW_AT_data_location         0x9733250effffffff000000001a12302928050033247e0022
                      DW_AT_sibling               <0x002d668a>
< 2><0x002d6509>      DW_TAG_member
                        DW_AT_name                  __hub__
                        DW_AT_type                  <0x002d64e0>
                        DW_AT_data_member_location  DW_OP_plus_uconst 0
< 2><0x002d6515>      DW_TAG_member
```
As you can see here we have for `DW_AT_data_location` a different dwarf-expression program
`0x9733250effffffff000000001a12302928050033247e0022`. The reason is that here it also takes the masking of the lowest 3 bits into account.

You should be able to use the same approach so that users do not need to know about the masking when they deref the hub field.

Reply:
Yes, this version of the hubname command is a hangover from an earlier version when I was omitting data_location attributes in the case where -H:-SpawnIsolates was specified. In the current version the mask should not be needed.

Reply:
The command can actually be simplified to this

    command hubname
      x/s ((_objhdr *)($arg0))->hub->name->value->data
    end

Reply:
Done